### ISSUE TITLE
Modelica-Buildings Library (master): Solver Failed to find spawn executable in Buildings Library installation

### ISSUE THREAD

**Issue by deepak19015** [OpenModelica/OpenModelica#11016]
### Description

The solver fails to find spawn executable in Buildings Library installation, on SPAWNPATH and on PATH. 
The following below listed models have same issues:

```
Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring
Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSummer
Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings
Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalTwoZoneBuildings
```
On simulating the model in OpenModelica leads to following errors/warnings:

```
1. Failed to find spawn executable in Buildings Library installation, on SPAWNPATH and on PATH. See installation instructions at Buildings.ThermalZones.EnergyPlus_9_6_0.UsersGuide.Installation
2. nonlinear system 1948 fails: at t=1.6848e+07
3. proper start-values for some of the following iteration variables might help [1] Real flo.building.isSynchronized(start=0,  nominal=1)
4. Solving non-linear system 1948 failed at time=16848000. For more information please use -lv LOG_NLS.
5. Failed to solve the initialization problem with global homotopy with equidistant step size.
6. Unable to solve initialization problem.
7. simulation terminated by an assertion at initialization
8. Simulation process failed. Exited with code 255.
```
The log report for these models are as given below: (Source: https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring.sim)

```
LOG_SUCCESS       | info    | The initialization finished successfully with 3 homotopy steps.
LOG_ASSERT        | info    | [/home/hudson/saved_omc/libraries/.openmodelica/libraries/Buildings 10.0.0-master/Media/Air.mo:146:5-147:57:writable]
|                 | |       | The following assertion has been violated at time 7347715.200000
|                 | |       | ((noEvent(flo.att.vol.medium.T >= 200.0))) --> "In IdealHeatingCoolingSpring.flo.att.vol.medium: Temperature T exceeded its minimum allowed value of -73.15 degC (200 Kelvin)
|                 | |       | as required from medium model "Buildings.Media.Air"."
LOG_ASSERT        | error   | No event found, but assert was triggered. Throwing now!
```
The list of warnings reported in the simulation test are given below: (Source: https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring.err)

```
1. The model contains alias variables with redundant start and/or conflicting nominal values. It is recommended to resolve the 
   conflicts, because otherwise the system could be hard to solve. To print the conflicting alias sets and the chosen candidates 
   please use -d=aliasConflicts.
```

### Steps to Reproduce

Load the Modelica-Buildings master library (Source: https://github.com/lbl-srg/modelica-buildings/)
Simulate the examples model present in Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring

### Expected Behavior

The simulation should succeed.

### Version and OS

- OpenModelica Version: OpenModelica 1.22.0~dev-250-g48ea6c4
- OS: Ubuntu 20.04 focal
      Versions of used Modelica libraries:
      Modelica 4.0.0
      Buildings master (10.0.0)

**Comment by AndreaBartolini** [comment]
Spawn executable shall be present in `path_to_library\Buildings\Buildings\Resources\bin\Spawn-light-0.4.3-7048a72798`, if not they shall be manually installed.

After installation, all the following models start to simulate but then they fail without error messages.
```
Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring
Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSummer
Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalTwoZoneBuildings
```
The model `Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings` fails because the file `modelica://Buildings/Resources/Data/ThermalZones/EnergyPlus_9_6_0/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance.idf`  is not found.

Under windows this can be related to the known problem of the long path management.

**Comment by casella** [comment]
@Pavanraj, I think this specific ticket can be closed. The CI is configured to install spawn now, and those models fail because of other issues that are mentioned in #13253.

Regarding the installation of spawn executables to run EnergyPlus models, also see lbl-srg/modelica-buildings#4055

**Comment by Pavanrajhr** [comment]
@casella @AndreaBartolini  In the current build, the models `Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSpring` and `Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.IdealHeatingCoolingSummer` 
are passing but the models 
`Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.Guideline36Winter`, `Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings` and `Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalTwoZoneBuildings` are failing. 
Is it still the issue with spawn installation still needs to be checked.

**Comment by casella** [comment]
> but the models `Buildings.ThermalZones.EnergyPlus_9_6_0.Examples.SmallOffice.Guideline36Winter`, `Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings` and `Buildings.ThermalZones.EnergyPlus_9_6_0.Validation.MultipleBuildings.TwoIdenticalTwoZoneBuildings` are failing. Is it still the issue with spawn installation still needs to be checked.

I guess you mean they are failing on your PCs, right?

On the CI, the first runs successfully and the other two also complete the simulation, but then generate an error because the spawn executable does not terminate correctly, see #13350 and lbl-srg/modelica-buildings#3345

Regarding the spawn installation instructions, @mwetter reported in [lbl-srg/modelica-buildings#4055](https://github.com/lbl-srg/modelica-buildings/issues/4055#issuecomment-2494121577) that he's updated the installation instructions.

@Pavanrajhr, @AndreaBartolini can you please try to make a clean install of spawn according to the latest instructions, without using python? We know the python script works in Linux because the CI uses it successfully, but on Windows it may not be available, so we should check the alternative ways of installing the binaries.

Thanks!

**Comment by JayHuLBL** [comment]
@casella We have opened a pull request to address the issue through MBL PR [#4585](https://github.com/lbl-srg/modelica-buildings/pull/4585). It seems to be that OpenModelica stores the external object pointer in its own runtime and then passes the same (now-freed) address again on the second destructor call.
We tried to set the pointers to NULL after being freed, as suggested https://cmu-sei.github.io/secure-coding-standards/sei-cert-c-coding-standard/recommendations/memory-management-mem/mem01-c/. However, by setting the pointer to NULL after `free()` cannot avoid the issue here.

**Comment by casella** [comment]
@AnHeuermann can you please check?

**Comment by AnHeuermann** [comment]
Sure, let's debug a bit.
Simulating `Buildings.ThermalZones.EnergyPlus_24_2_0.Examples.SmallOffice.IdealHeatingCoolingSpring` from Buildings master (+ running installation scripts in Buildings directory) works on my Ubuntu machine.

OMC is currently doing the following:

During initialization the initial system is solved:

1. OMC starts solving a non-linear system with `solve_nonlinear_system`.

      ```
      equation index: 2086
      indexNonlinear: 5
      type: NONLINEAR
      
      vars: {flo.building.isSynchronized}
      eqns: {2071, 2072, 2073, 2074, 2075, 2076, 2077, 2078, 2079, 2080, 2081, 2082, 2083, 2085, 2084}
      ```

2. Update inner equations: Evaluate residual function for non-linear system. So evaluate inner equation
    ```
    equation index: 2071
    type: SIMPLE_ASSIGN
    flo.sou.fmuZon.nObj = Buildings.ThermalZones.EnergyPlus_24_2_0.BaseClasses.initialize(flo.sou.fmuZon.adapter, flo.building.isSynchronized)
    ```

    This will call C function `initialize_Modelica_EnergyPlus_24_2_0` from Buildings/EnergyPlus_24_2_0_Wrapper.c and store the pointer to the initialized external object in `data->simulationInfo->extObjs[6]`.

In a similar fashion other non-linear systems get solved an will initialize `data->simulationInfo->extObjs[7]`, `data->simulationInfo->extObjs[0]`, `data->simulationInfo->extObjs[3]`, `data->simulationInfo->extObjs[5]` and `data->simulationInfo->extObjs[4]`.

Then the actual simulation system is solved during the end of the initialization. Here the non-linear systems from above are solved again, so the external objects in 6, 7, 0, 3, 5 and 4 get initialized again. That's probably not good.

After the simulation is done we call all destructors on external objects 6,4,5,7,3,0.

_______________________________

Bottom line: 

* OMC initializes the external objects twice (probably bad, but should "only" result in memory leaks)
* The destructor function is called only once at the end

So, the main issue is: Why does OMC put the external object constructor equations in  the inner equations of the non-linear systems?

**Comment by AnHeuermann** [comment]
For the failing `Buildings.ThermalZones.EnergyPlus_24_2_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings` we are initializing an external object in `data->simulationInfo->extObjs[4]` and `data->simulationInfo->extObjs[5]` (a bunch of times, again because it is in the non-linear system). When calling the desctructor on `data->simulationInfo->extObjs[5]` the executable crashes. For some reason both `data->simulationInfo->extObjs[4]` and `data->simulationInfo->extObjs[5]` point to the exact same pointer address during the cleaning step which is the root cause for the double free.

External object 5 is initialized in 

```
/*
equation index: 247
type: SIMPLE_ASSIGN
bui[2].zon.fmuZon.adapter = Buildings.ThermalZones.EnergyPlus_24_2_0.BaseClasses.SpawnExternalObject.constructor(1, bui[2].zon.fmuZon.startTime, "TwoIdenticalOneZoneBuildings.bui.building", "TwoIdenticalOneZoneBuildings.bui.zon", "spawn-0.6.0-9f1b36b00b", "24_2_0", bui[2].zon.fmuZon.idfName, bui[2].zon.fmuZon.epwName, bui[2].zon.fmuZon.runPeriod, bui[2].zon.fmuZon.relativeSurfaceTolerance, bui[2].zon.fmuZon.zoneName, false, "", "/home/andreas/workdir/modelica-libraries/modelica-buildings/Buildings/legal.html", bui[2].zon.fmuZon.logLevel, false, "zones", "        \"name\": \"" + bui[2].zon.fmuZon.zoneName + "\"", {"AFlo", "V", "mSenFac"}, {"m2", "m3", "1"}, 3, {"T", "X", "mInlets_flow", "TAveInlet", "QGaiRad_flow"}, {"K", "1", "kg/s", "K", "W"}, 5, {"TRad", "QConSen_flow", "QLat_flow", "QPeo_flow"}, {"K", "W", "W", "W"}, 4, {}, 0, {})
*/
```

`allocate_Spawn_EnergyPlus_24_2_0` is returning the same pointer it did return for the other `data->simulationInfo->extObjs[4]`. My best guess is because it is using the same `modelicaName`. OMC isn't aware that pointers are re-used and expects the two constructor calls to be independent of each other.

**Comment by AnHeuermann** [comment]
I double-checked and indeed the issue for OMC is that external function `allocate_Modelica_EnergyPlus_24_2_0` will return the exact same pointer (e.g., `0x5bbbf9c6fb60`) for the two `bui[1].zon.fmuZon.adapter` and `bui[2].zon.fmuZon.adapter`:

```
[allocate_Modelica_EnergyPlus_24_2_0] IN: objectType=1 startTime=0 modelicaNameBuilding=TwoIdenticalOneZoneBuildings.bui.building modelicaInstanceName=TwoIdenticalOneZoneBuildings.bui.zon spawnExe=spawn-0.6.0-9f1b36b00b idfVersion=24_2_0 idfName=/path/to/Buildings/Resources/Data/ThermalZones/EnergyPlus_24_2_0/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance.idf epwName=/path/to/Buildings/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw runPeriod=(startDayOfYear=7 applyWeekEndHolidayRule=0 use_weatherFileDaylightSavingPeriod=0 use_weatherFileHolidaysAndSpecialDays=0 use_weatherFileRainIndicators=1 use_weatherFileSnowIndicators=1) relativeSurfaceTolerance=1e-06 epName=LIVING ZONE usePrecompiledFMU=0 fmuName= buildingsRootFileLocation=/path/to/Buildings/legal.html logLevel=2 printUnit=0 jsonName=zones jsonKeysValues=        "name": "LIVING ZONE" nParOut=3 nInp=5 nOut=4 nDer=0
[allocate_Modelica_EnergyPlus_24_2_0] OUT: adapter_ext=0x5bbbf9c6fb60
EnergyPlus Starting
EnergyPlus, Version 24.2.0-9f1b36b00b, YMD=2026.05.19 13:49
Adjusting Air System Sizing
Adjusting Standard 62.1 Ventilation Sizing
Initializing Simulation
Reporting Surfaces
Beginning Primary Simulation
Initializing New Environment Parameters
Warming up {1}
Warming up {2}
Warming up {3}
Warming up {4}
Warming up {5}
Warming up {6}
Warming up {7}
Warming up {8}
Warming up {9}
Warming up {10}
Starting Simulation at 01/01/2006 for SPAWN-RUNPERIOD
[allocate_Modelica_EnergyPlus_24_2_0] IN: objectType=1 startTime=0 modelicaNameBuilding=TwoIdenticalOneZoneBuildings.bui.building modelicaInstanceName=TwoIdenticalOneZoneBuildings.bui.zon spawnExe=spawn-0.6.0-9f1b36b00b idfVersion=24_2_0 idfName=/path/to/Buildings/Resources/Data/ThermalZones/EnergyPlus_24_2_0/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance.idf epwName=/path/to/Buildings/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw runPeriod=(startDayOfYear=7 applyWeekEndHolidayRule=0 use_weatherFileDaylightSavingPeriod=0 use_weatherFileHolidaysAndSpecialDays=0 use_weatherFileRainIndicators=1 use_weatherFileSnowIndicators=1) relativeSurfaceTolerance=1e-06 epName=LIVING ZONE usePrecompiledFMU=0 fmuName= buildingsRootFileLocation=/path/to/Buildings/legal.html logLevel=2 printUnit=0 jsonName=zones jsonKeysValues=        "name": "LIVING ZONE" nParOut=3 nInp=5 nOut=4 nDer=0
[allocate_Modelica_EnergyPlus_24_2_0] OUT: adapter_ext=0x5bbbf9c6fb60
```

While it can be argued that it would be better if OMC would optimize and only instantiate one external object I think it is still compliant with the Modelica specification to initialize two external objects that are exactly the same. And in that case to call the external destructor for both objects.

@JayHuLBL Maybe it is feasible to add logic to `free_Spawn_EnergyPlus_24_2_0` destructor to save which external objects were already freed. I assume you save the pointer during the constructor call already to be able to skip initializing the same thing twice. 

```C
  /* Dymola 2019FD01 calls in some cases the allocator twice. In this case, simply return the previously instanciated Spawn object pointer */
  setExchangePointerIfAlreadyInstanciated(modelicaName, objectType, &ptrSpaObj);
```
In that case you could also keep track of the ones deleted already and skip destructor calls in that case.

**Comment by JayHuLBL** [comment]
@AnHeuermann The refactored `free_Spawn_EnergyPlus_24_2_0()` (file changes in https://github.com/lbl-srg/modelica-buildings/pull/4585) already tracks which objects have been freed (implicitly, via their absence from the `exchange[]` arrays) and skips destructor calls for already-freed objects.
Adding a separate static freed-objects list would introduce a second mutable registry that have to be kept in sync, and would require a new clearing lifecycle. Would it be redundant?

**Comment by AnHeuermann** [comment]
> @AnHeuermann The refactored free_Spawn_EnergyPlus_24_2_0() (file changes in https://github.com/lbl-srg/modelica-buildings/pull/4585) already tracks which objects have been freed (implicitly, via their absence from the exchange[] arrays) and skips destructor calls for already-freed objects.

Sounds like this should be enough. Basically OpenModelica does the following

```C
void *ptr1 = allocate_Modelica_EnergyPlus_24_2_0(/* building[1] */);
void *ptr2 = allocate_Modelica_EnergyPlus_24_2_0(/* building[2] */); 

// Since the buildings are equal we have ptr1 = ptr2

/* Run simulation */

free_Spawn_EnergyPlus_24_2_0(prt1);
free_Spawn_EnergyPlus_24_2_0(prt2);
```

As long as this is working the simulation should pass.

Did you test if your proposed changes succeed for the models?

**Comment by JayHuLBL** [comment]
@AnHeuermann 
With OpenModelica verson 1.27.0~dev-526-g4bff1ce, following models:
```
Buildings.ThermalZones.EnergyPlus_24_2_0.Examples.SmallOffice.IdealHeatingCoolingSpring
Buildings.ThermalZones.EnergyPlus_24_2_0.Examples.SmallOffice.IdealHeatingCoolingSummer
Buildings.ThermalZones.EnergyPlus_24_2_0.Validation.MultipleBuildings.TwoIdenticalOneZoneBuildings
Buildings.ThermalZones.EnergyPlus_24_2_0.Validation.MultipleBuildings.TwoIdenticalTwoZoneBuildings
```
run without issue.

**Comment by casella** [comment]
Great!

@NehaSPatwardhan please update the Excel file accordingly.
